### DISCUSSION TITLE
Running Phase4 on multiple server instances

### DISCUSSION THREAD

**Discussion by tzkmn** [phax/phase4#112] [Q&A]
Hello.

We are considering using a Phase4 implementation on an auto-scaling infrastructure such as Amazon Web Services or Google Cloud Platform.
Therefore, Phase4 will run on multiple server instances.
However, in a Phase4 implementation, there are some managers that manage state on each server instance, such as AS4DuplicateManager.
For example, if the sender's AP performs a retry process due to some problem, a message with the same messageID may be received on different server instances.
In this case, AS4DuplicateManager may not be able to detect the duplicate messageID.

Could you tell me how to deploy Phase4 implementation on multiple server instances if you have any ideas?
Thank you very much for your support.

Best Regards,
Takayuki Tsuzuki

**Comment by phax** [comment]
Dear Tsuzuki-san,
Indeed, using the `AS4DuplicateManager` as-is will not work.
To have an auto-scaling solution to this problem, you need to provide your own implementation of `IAS4DuplicateManager` that e.g. works on a database or something else.
See https://github.com/phax/phase4/blob/master/phase4-peppol-client/src/test/java/com/helger/phase4/peppol/supplementary/issues/MainDiscussion112.java for an example how to set a custom duplicate manager.

hth, Philip

**Reply by tzkmn** [reply]
Dear Philip-san,
Happy new year and Thank you very much for your answer.  
I will try to create and apply my own implementation.

Best Regards,
Takayuki Tsuzuki

**Reply by phax** [reply]
Dear Tsuzuki-san,
If you come up with a solution that you want to share with the community, I am happy to integrate it into the code base. The only requirement I have, is that you would agree to share your code under the existing Apache 2 License.
Thanks, Philip
